src/partition/repart.c

```diff
@@ -479,6 +480,42 @@ static uint64_t round_up_size(uint64_t v, uint64_t p) {
         return v * p;
 }
 
+/* calculates the size of a dm-verity hash partition's contents */
+static int calculate_verity_hash_size(
+                uint64_t data_bytes,
+                uint64_t hash_block_size,
+                uint64_t data_block_size,
+                uint64_t *ret_bytes) {
+
+        uint64_t data_blocks = DIV_ROUND_UP(data_bytes, data_block_size);
+        if (data_blocks > UINT64_MAX / data_block_size)
+                return -EOVERFLOW;
+
+        /* hashes that fit in one hash block (node in the merkle tree) */
+        uint64_t hashes_per_hash_block = hash_block_size / SHA256_DIGEST_SIZE;
+
+        /* initialize with 1 for the root of the merkle tree */
+        uint64_t hash_blocks = 1;
+
+        /* iterate through the levels of the merkle tree bottom up */
+        uint64_t remaining_blocks = data_blocks;
+
+        while (remaining_blocks > hashes_per_hash_block) {
+                uint64_t hash_blocks_for_level;
+                /* number of hash blocks required to reference the underlying blocks */
+                hash_blocks_for_level = DIV_ROUND_UP(remaining_blocks, hashes_per_hash_block);
+                /* add current layer to the total number of hash blocks */
+                hash_blocks += hash_blocks_for_level;
+                /* hashes on this level serve as the blocks on which the next level is built */
+                remaining_blocks = hash_blocks_for_level;
+        }
+
+        /* add one for the superblock */
+        *ret_bytes = (hash_blocks + 1) * hash_block_size;
+
```

Comment:
where is this algorithm taken from? is this documented somewhere? or is this a reimplementation of similar code in veritysetup or so?

Can you please add comments referencing the source and the upstream definition of this logic?

Reply:
Upstream calculates the size while generating the merkle tree, this was implemented from scratch, originally in python.
Do you think referencing docs for the on-disk layout is enough here?
https://docs.kernel.org/admin-guide/device-mapper/verity.html#hash-tree
https://gitlab.com/cryptsetup/cryptsetup/-/wikis/DMVerity#hash-tree

Reply:
This should be the relevant upstream code:

1. calculation of the nr. of levels in the tree: 
https://gitlab.com/cryptsetup/cryptsetup/-/blob/main/lib/verity/verity_hash.c?ref_type=heads#L90
2. calculation of the hash tree size:
https://gitlab.com/cryptsetup/cryptsetup/-/blob/main/lib/verity/verity_hash.c?ref_type=heads#L297-L302

Reply:
all i care about is that the interested reader can quickly review this by finding where this comes from. After all this logic is not defined by us, but by the kernel/veritysetup, we just do a really limited reimplementation of one facet of it. And because of these roles, we really should reference who actually defines this. A bunch of links would be enough.

Reply:
added a comment referencing both kernel docs and the relevant upstream code